types/src/jsonrpc/wrapped/batches.rs

```diff
@@ -180,32 +178,18 @@ impl<T> BatchesState<T> {
 	/// Injects a newly-received batch into the list. You must then call
 	/// [`next_event`](BatchesState::next_event) in order to process it.
 	pub fn inject(&mut self, request: jsonrpc::Request, user_param: T) {
+		let batch_id = self.vacant.pop().unwrap_or_else(|| self.batches.len());
 		let batch = batch::BatchState::from_request(request);
 
-		loop {
-			let id = self.next_batch_id;
-			self.next_batch_id = self.next_batch_id.wrapping_add(1);
-
-			// We shrink `self.batches` from time to time so that it doesn't grow too much.
-			if id % 256 == 0 {
-				self.batches.shrink_to_fit();
-				// TODO: self.batches.shrink_to(BATCHES_MIN_CAPACITY);
-				// ^ see https://github.com/rust-lang/rust/issues/56431
-			}
-
-			match self.batches.entry(id) {
-				Entry::Occupied(_) => continue,
-				Entry::Vacant(e) => {
-					e.insert((batch, user_param));
-					break;
-				}
-			}
-		}
+		self.batches.insert(batch_id, Some((batch, user_param)));
 	}
 
 	/// Returns a list of all user data associated to active batches.
 	pub fn batches<'a>(&'a mut self) -> impl Iterator<Item = &'a mut T> + 'a {
-		self.batches.values_mut().map(|(_, user_data)| user_data)
+		self.batches.iter_mut().filter_map(|entry| match entry {
+			Some((_, user_param)) => Some(user_param),
+			None => None,
+		})
```

Comment:
Could this be `self.batches.iter_mut().filter_map(|entry| entry.map(|(_, user_param)| user_param ))` or something similar?